### PR TITLE
Add `live` progress style and refactor output and progress styles

### PR DESCRIPTION
**Pull request by JoshKarpel**
Putting this up for some early feedback. Have not tested all the corner cases of #205 yet.

@darrenburns I did what I described in #227 and it works great! I think I came up with a fairly clean way of abstracting away the actual test running while giving the styles a fair bit of control over what the output looks like. I've got the "live" output style you describe in the issue, plus a "none" output style that prints nothing, PLUS dots work with progress bars! Some of these are even a little faster in wall clock time because more of the printing is happening on a background thread.

... but none of them will work in CI except test-per-line output with inline progress, because the others all use `Live`s. In fact, since I'm always starting the live right now, it may mess up CI even with the components that would use it turned off....

### CODE REVIEW

**Comment by darrenburns** [comment]
Looks really cool! Big fan of how you've refactored the output styles code too, it feels _much_ cleaner and seems like it'll make it much easier to offer a hook for plugins to offer custom output styles 🙂 

From a UX point of view, I think we'd need a way of seeing if any tests have failed so far (maybe having the little "Results" panel that displays after a test run always be visible and live updating? - not sure if that's trying to stretch the Live API too far, although I haven't looked enough at it)

An issue though, that I suspect may throw a huge spanner in the works here but would _love_ to be wrong... is that it breaks debugging. It seems like the `Live` continues trying to update the terminal while you're inside `pdb` 😬 We would probably need a method of suspending running `Live` inside [this function](https://github.com/darrenburns/ward/blob/936d00edc230ff3cdc8b0e5d229c65099bbe16e9/ward/_debug.py#L22). Whether that's even possible since it'd be outside the context of the context manager 🤷

**Comment by JoshKarpel** [comment]
> From a UX point of view, I think we'd need a way of seeing if any tests have failed so far (maybe having the little "Results" panel that displays after a test run always be visible and live updating? - not sure if that's trying to stretch the Live API too far, although I haven't looked enough at it)

I was actually thinking of something similar - the failed tests could get printed, so they would be "left behind" by the single updating line. I just pushed this up for you to take a look.

> An issue though, that I suspect may throw a huge spanner in the works here but would _love_ to be wrong... is that it breaks debugging. It seems like the `Live` continues trying to update the terminal while you're inside `pdb` 😬 We would probably need a method of suspending running `Live` inside [this function](https://github.com/darrenburns/ward/blob/936d00edc230ff3cdc8b0e5d229c65099bbe16e9/ward/_debug.py#L22). Whether that's even possible since it'd be outside the context of the context manager 🤷

I actually did exactly this for Spiel so that I could drop into a terminal session https://github.com/JoshKarpel/spiel/blob/b37497490dd12e693d481e0c5d81a797c7b922c5/spiel/input.py#L386

As long as you can get a reference to the `Live`, you're good to go. I'll see if I can figure out a sane way to get a reference over there...

**Comment by JoshKarpel** [comment]
Bah, I over-promised. Turns out you can't really wrap the breakpoint call in a context manager. I went with the next-easiest option, which is to disable the `Live` when entering the breakpoint and unconditionally make sure it's running after each test finishes. Kind of weird behavior if you `n` off the end of the test in the debugger all the way into Ward's internals.

**Comment by JoshKarpel** [comment]
Ok, I think this is ready for a round of closer review.

@darrenburns despite the refactor, I'm still not really sure how to test these in a clean way. I really like how the classes manage their own state, but they're big bags of complicated state now, and it's hard to drive them in a consistent way without executing a test suite... which you can't do because the test suite itself is running a `Live` and you can't have two `Live`s running at once. I'm tempted to defer intensive testing of the internals of this once again, and hope that in some other PR we can build a way to test that that is robust and simple enough to not need continual tweaking as the terminal code evolves.

**Comment by darrenburns** [comment]
@JoshKarpel I'll have a look at this soon, hopefully within the next couple of days.

> I think we'll have to ditch the enumerated styles to make this work, if it involves adding new widget classes and registering them somewhere :(

That's likely, yeah. There might be a nice `__init_subclass__` approach, but will need to think more.

File: tests/test_terminal.py [hunk 1]
```diff
@@ -206,3 +210,93 @@ def _(timing_stats_panel=timing_stats_panel):
...
+                    fn=lambda: 1,
+                    description="test1",
+                    module_name="mod1",
+                ),
+                outcome=outcome,
+            )
+        )
+
+
+@test(f"{LiveTestBar.__name__}'s emoji changes the first time a test fails")
```
Comment (darrenburns) [inline tests/test_terminal.py:283]: As much as I love an emoji, I think I preferred the output without the emojis, sorry 😢 Hopefully this sort of thing will be possible via plugins soon!

File: ward/_terminal.py [hunk 2]
```diff
@@ -504,17 +311,359 @@ def handle_result(self, test_result: TestResult):
...
+    def after_test(self, test_index: int, test_result: TestResult) -> None:
+        self.progress.update(self.task, advance=1)
+        self.test_description_column.renderable = get_test_result_line(
+            test_result=test_result,
+            test_index=test_index,
+            num_tests=self.num_tests,
+            progress_styles=self.progress_styles,
+        )
+
+        if test_result.outcome is TestOutcome.FAIL:
```
Comment (darrenburns) [inline ward/_terminal.py:530]: We should also leave any unexpected passes (XPASS) we encounter.

**Review by darrenburns — CHANGES_REQUESTED** [review]
Won't be able to look any more for now, but left a couple of comments.

I've got to say in general I absolutely love how clean the code looks now 👏

File: tests/test_terminal.py (hunk 1 above)
Comment (JoshKarpel) [reply tests/test_terminal.py:283]: 😆 no worries! I was mostly playing with ways of spicing up the output.

File: ward/testing.py [hunk 3]
```diff
@@ -472,6 +472,10 @@ def display_name(self):
         assert len(display_names) == len(TestOutcome)
         return display_names[self]
 
+    @property
+    def is_bad(self) -> bool:
```
Comment (JoshKarpel) [inline ward/testing.py:476]: @darrenburns any thoughts on what to call "a test outcome that represents a failure but may not literally be `FAIL`"?

Comment (darrenburns) [reply ward/testing.py:476]: Some ideas:
* `is_fail_or_xpass` - Explicit, although there is a possibility that there could be an `ERR` test outcome too in the future.
* `causes_session_failure` - We could categorise outcomes in outcomes that cause the test suite to fail, and those that don't.  The occurrence of a single `XPASS` or a `FAIL` in a suite causes the overall test session to end in a failure.

File: ward/testing.py [hunk 4]
```diff
@@ -472,6 +472,14 @@ def display_name(self):
         assert len(display_names) == len(TestOutcome)
         return display_names[self]
 
+    @property
+    def will_fail_session(self) -> bool:
```
Comment (darrenburns) [inline ward/testing.py:476]: Could we add tests around the two properties that were added here?

Comment (JoshKarpel) [reply ward/testing.py:476]: Done in https://github.com/darrenburns/ward/pull/233/commits/fc864cc344ca569f1e6bc641c49960ff234eb338

**Review by darrenburns — APPROVED** [review]

**Comment by darrenburns** [comment]
This is great, thanks! ❤️
